### PR TITLE
Add a scirpt to report broken links in all markdowns.

### PR DESCRIPTION
**Pull request by HarshTrivedi**
This PR adds a script to report broken links in all markdowns in the project.

```
time python scripts/verify.py --checks check-links
UnReachable Links:
Verifying with ['check-links']
LinksValidity (check):

UnReachable Links:

0.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: on GitHub
Link: https://github.com/allenai/allennlp-as-a-library-example/tree/0.5.1

1.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: `static_html`
Link: https://github.com/allenai/allennlp-as-a-library-example/tree/0.5.1/static_html

2.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: its own file
Link: https://github.com/allenai/allennlp-as-a-library-example/blob/0.5.1/static_html/demo.css

3.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: add a `script` tag to load chart.js
Link: https://github.com/allenai/allennlp-as-a-library-example/blob/0.5.1/static_html/index.html#L47

4.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: `my_library/__init__.py`
Link: https://github.com/allenai/allennlp-as-a-library-example/blob/0.5.1/my_library/__init__.py

5.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: can be very simple
Link: https://github.com/allenai/allennlp-as-a-library-example/blob/0.5.1/my_library/predictors/paper_classifier_predictor.py

6.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: placing that inside our `output` div
Link: https://github.com/allenai/allennlp-as-a-library-example/blob/0.5.1/static_html/index.html#L61

7.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_pretrained_models.md
Name: Named Entity Recognition model
Link: http://allennlp.org/models/#named-entity-recognition

8.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: as a test fixture
Link: https://github.com/allenai/allennlp-as-a-library-example/tree/0.5.1/tests/fixtures

9.
Source: /Users/harshtrivedi/allennlp/tutorials/getting_started/using_as_a_library_pt2.md
Name: `text_to_instance`
Link: https://github.com/allenai/allennlp-as-a-library-example/blob/0.5.1/my_library/dataset_readers/semantic_scholar_papers.py#L68

python scripts/verify.py --checks check-links  2.97s user 0.70s system 34% cpu 10.761 total
```

Takes about 10 seconds.

### CODE REVIEW

File: scripts/check_links.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+"""
+import re
+import sys
+import pathlib
+from multiprocessing.dummy import Pool
+from collections import namedtuple
+from typing import Tuple
+import requests
+
+MatchTuple = namedtuple('MatchTuple', 'source name link')
```
Comment (DeNeutoy) [inline scripts/check_links.py]: Did you know that you can create `NamedTuple` classes by inheriting from `typing.NamedTuple`? It allows you to have types in the tuple, which is nice - see [here](https://github.com/allenai/allennlp/blob/master/allennlp/models/constituency_parser.py#L21) for an example.

File: scripts/check_links.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    url_regex = re.compile(r'\[([^!][^\]]+)\]\((http[s]?[^)(]+)\)')
+    for markdown_file in markdown_files:
+        with open(markdown_file) as handle:
+            for line in handle.readlines():
+                matches = url_regex.findall(line)
+                for match in matches:
+                    if 'localhost' not in match[1]:
+                        all_matches.add(MatchTuple(source=markdown_file, name=match[0], link=match[1]))
+
+    with Pool(processes=10) as pool:
```
Comment (DeNeutoy) [inline scripts/check_links.py]: How slow is this without multiprocessing? I'd probably prefer without, unless the speedup is significant.

File: scripts/check_links.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        RESULTS = pool.map(url_ok, [match for match in list(all_matches)])
```
Comment (DeNeutoy) [inline scripts/check_links.py]: results vs `RESULTS`?

File: scripts/verify.py [hunk 4]
```diff
@@ -37,12 +37,17 @@ def main(checks):
             run("./scripts/check_docs.py", shell=True, check=True)
             print("check docs passed")
 
+        if "check-links" in checks:
+            print("LinksValidity (check):", flush=True)
```
Comment (DeNeutoy) [inline scripts/verify.py]: "Checking links in Markdown files" might be a better comment.

**Comment by HarshTrivedi** [comment]
Sure 👍, can you tell me how to add it to CI?

File: scripts/check_links.py (hunk 3 above)
Comment (HarshTrivedi) [reply scripts/check_links.py]: Ohh, missed reverting that. my local pylint complained but then found out checks aren't being done on these scripts.

**Comment by nelson-liu** [comment]
speaking of which: is there any easy way teamcity can read its config from the repo (à la travis)? This would make it a lot easier for people to (1) see what's going on in the CI if they want to find out and (2) enable people to contribute to it.

File: scripts/verify.py (hunk 4 above)
Comment (HarshTrivedi) [reply scripts/verify.py]: Cool. Actually, I was trying to match the style of other items like `Documentation (build):`, `Linter (pylint):` .. etc etc.

File: scripts/check_links.py (hunk 1 above)
Comment (HarshTrivedi) [reply scripts/check_links.py]: Cool, I didn't know. Let me check ...

File: scripts/check_links.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    project_root = (pathlib.Path(__file__).parent / "..").resolve() # pylint: disable=no-member
+    markdown_files = project_root.glob('**/*.md')
+
+    all_matches = set()
+    url_regex = re.compile(r'\[([^!][^\]]+)\]\((http[s]?[^)(]+)\)')
+    for markdown_file in markdown_files:
+        with open(markdown_file) as handle:
+            for line in handle.readlines():
+                matches = url_regex.findall(line)
+                for match in matches:
```
Comment (joelgrus) [inline scripts/check_links.py]: this would be a little bit cleaner if you did

```
for name, link in matches:
```

Comment (HarshTrivedi) [reply scripts/check_links.py]: Sure 👍 Thanks

**Comment by joelgrus** [comment]
it would also be helpful if it could check for stale links; that is, links to a specific tag that's not the last released version, although I don't know if it's easy to programmatically get the last released version

File: scripts/check_links.py (hunk 2 above)
Comment (HarshTrivedi) [reply scripts/check_links.py]: yeah its' about 5X. With parallel about 8-12 seconds. Without parallel 40-60 seconds

**Comment by HarshTrivedi** [comment]
@joelgrus  Do you mean if following link is found: `https://github.com/allenai/allennlp-as-a-library-example/tree/0.4.0` with tag`0.4.0`, even though it is working it should report? If not, can you give an example?

**Comment by schmmd** [comment]
@nelson-liu unfortunately there's no easy way to do that AFAIK.  Part of the reason is likely because TC's configuration is much more complex than Travis'.

**Comment by schmmd** [comment]
@DeNeutoy easy to add to TC.  You should get this merged first!

**Comment by joelgrus** [comment]
@HarshTrivedi yes, that's what I meant, but really I should make that check part of our release process, so never mind

**Pull request merged**